### PR TITLE
feat(coderd/notifications): improve notification format consistency

### PR DESCRIPTION
**Pull request by SasSwart**
This Pull request addresses the more trivial items in https://github.com/coder/coder/issues/14893.
These were simple formatting changes that I was able to fix despite limited context.

Some more changes are required for which I will have to dig a bit deeper into how the template contexts are populated. I'm happy to add those to this PR or create a subsequent PR.

### CODE REVIEW

**Comment by SasSwart** [comment]
I have read the CLA Document and I hereby sign the CLA

File: coderd/database/migrations/000261_improve_notification_templates.up.sql
Comment (mafredri) [inline coderd/database/migrations/000261_improve_notification_templates.up.sql:5]: Have we allowed any customers to manually change these templates or mentioned it's a supported use-case? If yes, we should probably validate that the template hasn't been changed before updating it?

//cc: @dannykopping @mtojek

Comment (mtojek) [reply coderd/database/migrations/000261_improve_notification_templates.up.sql:5]: Good question, but currently it is not supported 👍

**Review by mtojek — COMMENTED** [review]
I think you need to regenerate the text samples with `make update-golden-files`, or overwrite default for `updateGoldenFiles` (false -> true) and rerun test.

Once we have new samples, it will be straightforward to review changes 👍

**Comment by mafredri** [comment]
> I think you need to regenerate the text samples with `make update-golden-files`, or overwrite default for `updateGoldenFiles` (false -> true) and rerun test.

We should make the test output this hint on error (i.e. change requires to assert and if test failed, show hint). Right now it's only shown when there are missing files.

File: coderd/database/migrations/000261_improve_notification_templates.up.sql
Comment (SasSwart) [reply coderd/database/migrations/000261_improve_notification_templates.up.sql:5]: I did find the way we currently set and update these templates to be unintuitive and hard to work with. I have an idea that would make it easier to track changes to templates, but I'm not sure if it will be worth the complexity.

What I have in mind would introduce an auto incrementing "revision" column to the notification_templates table. This would make up and down migrations that adjust templates simpler, but it would make the queries that get the template in order to render it slightly more complex.

**Comment by SasSwart** [comment]
> I think you need to regenerate the text samples with make update-golden-files, or overwrite default for updateGoldenFiles (false -> true) and rerun test.

I was trying to do this for most of Friday. Turns out that the makefile wasn't actually calling the right test to update these. Firstly because it didn't have a job in the makefile for the notifications package, but also because the test wasn't named correctly.

I've pushed a commit that fixes it. We now have updated testdata, which has shown some issues. I'm working on resolving them. Expect updated queries and rendered testdata soon.

File: coderd/notifications/notifications_test.go
Comment (mafredri) [inline coderd/notifications/notifications_test.go:971]: Sprintf is missing in these two.

**Review by mafredri — APPROVED** [review]
Except for missing fmt, these look good to me. Would like to see what @mtojek thinks before merging though.

**Review by mtojek — APPROVED** [review]
Definitely better than it is now! Thanks 👍

**Comment by SasSwart** [comment]
@mtojek @mafredri 

Apologies, but I wasn't happy with the standard of the PR. I reworked it after considering some verbal feedback (thanks @mafredri and @defelmnq).

I also figured out how to deal with the rest of #14893.

Please have another look?

Enough has changed that I would have revoked the initial approvals if I could.
I'd particularly appreciate your thoughts on my changes in functions postUser, postFirstUser and oauthLogin

File: coderd/database/migrations/000262_improve_notification_templates.up.sql
Comment (dannykopping) [inline coderd/database/migrations/000262_improve_notification_templates.up.sql:6]: Awesome, thanks for including the change description.

File: coderd/database/migrations/000262_improve_notification_templates.up.sql
Comment (dannykopping) [inline coderd/database/migrations/000262_improve_notification_templates.up.sql:64]:
```suggestion
	body_template = E'Hi {{.UserName}},\n\n' || -- Add a comma
```

File: coderd/notifications/testdata/rendered-templates/TemplateUserAccountCreated-body.md.golden
Comment (dannykopping) [inline coderd/notifications/testdata/rendered-templates/TemplateUserAccountCreated-body.md.golden:5]: This doesn't seem right.

File: coderd/notifications/testdata/rendered-templates/TemplateWorkspaceManualBuildFailed-body.md.golden
Comment (dannykopping) [inline coderd/notifications/testdata/rendered-templates/TemplateWorkspaceManualBuildFailed-body.md.golden:5]:
```suggestion
The template's display name was **William's Template**. The workspace build was initiated by **joe**.
```

**Review by dannykopping — COMMENTED** [review]
Noticed a few problems in the rendered templates.
Please go through these closely to make sure they are as you're expecting.

File: coderd/database/migrations/000262_improve_notification_templates.up.sql
Comment (SasSwart) [reply coderd/database/migrations/000262_improve_notification_templates.up.sql:6]: Well that's embarrassing... I might try letting a PR rest overnight so I can double check with fresh eyes before requesting review next time. Thanks for catching.

File: coderd/database/migrations/000262_improve_notification_templates.up.sql
Comment (SasSwart) [reply coderd/database/migrations/000262_improve_notification_templates.up.sql:64]: Well that's embarrassing... I might try letting a PR rest overnight so I can double check with fresh eyes before requesting review next time. Thanks for catching.

File: coderd/database/migrations/000261_improve_notification_templates.up.sql
Comment (SasSwart) [reply coderd/database/migrations/000261_improve_notification_templates.up.sql]: This should no longer be an issue, because we've moved away from using REPLACE

File: coderd/notifications/testdata/rendered-templates/TemplateWorkspaceDormant-body.md.golden
Comment (SasSwart) [inline coderd/notifications/testdata/rendered-templates/TemplateWorkspaceDormant-body.md.golden:3]: Note to self. Fix this

File: coderd/database/migrations/000262_improve_notification_templates.up.sql
Comment (SasSwart) [reply coderd/database/migrations/000262_improve_notification_templates.up.sql]: Fixed

File: coderd/notifications/testdata/rendered-templates/TemplateWorkspaceManualBuildFailed-body.md.golden
Comment (SasSwart) [reply coderd/notifications/testdata/rendered-templates/TemplateWorkspaceManualBuildFailed-body.md.golden]: Fixed

File: coderd/notifications/testdata/rendered-templates/TemplateUserAccountCreated-body.md.golden
Comment (SasSwart) [reply coderd/notifications/testdata/rendered-templates/TemplateUserAccountCreated-body.md.golden]: Fixed

File: coderd/users.go
Comment (SasSwart) [inline coderd/users.go:198]: @mafredri Are we happy to hardcode this value here?

File: coderd/database/migrations/000262_improve_notification_templates.down.sql
Comment (mtojek) [inline coderd/database/migrations/000262_improve_notification_templates.down.sql:17]: nit: I guess, a missed indent?

File: coderd/database/migrations/000262_improve_notification_templates.down.sql
Comment (mtojek) [inline coderd/database/migrations/000262_improve_notification_templates.down.sql:33]: nit: same here (indent)

File: coderd/notifications/notifications_test.go
Comment (mtojek) [inline coderd/notifications/notifications_test.go:975]: nit: could be const too

File: coderd/users.go
Comment (mtojek) [reply coderd/users.go:198]: Correct me if I'm wrong, in this case, "coder" indicates the _system_? If so, then 👍 

re hardcoding:

if this is the only place, then I believe it is fine.

File: coderd/notifications/testdata/rendered-templates/TemplateTemplateDeleted-body.md.golden
Comment (mtojek) [inline coderd/notifications/testdata/rendered-templates/TemplateTemplateDeleted-body.md.golden:5]: The display name is just a human-readable representation of the template name, so in this case, you don't need to indicate that it is not the real template name.

Just go with:

> The template **Bobby's Template** was deleted by **rob**.

File: coderd/notifications/testdata/rendered-templates/TemplateUserAccountActivated-body.md.golden
Comment (mtojek) [inline coderd/notifications/testdata/rendered-templates/TemplateUserAccountActivated-body.md.golden:5]: Have we considered a bit different phrase?

> The account for bobby has been activated by Rob. This account belongs to William Tables.

File: coderd/notifications/testdata/rendered-templates/TemplateYourAccountSuspended-body.md.golden
Comment (mtojek) [inline coderd/notifications/testdata/rendered-templates/TemplateYourAccountSuspended-body.md.golden:3]: nit: trailing dot

File: coderd/notifications/testdata/rendered-templates/TemplateUserAccountDeleted-body.md.golden
Comment (mtojek) [inline coderd/notifications/testdata/rendered-templates/TemplateUserAccountDeleted-body.md.golden:5]: Curious, if we can make notifications `TemplateUserAccountDeleted`, `TemplateUserAccountCreated`, `TemplateUserAccountActivated` and `TemplateUserAccountSuspended` sound more natural?

Any ideas, @dannykopping? Should we involve Christin?

Comment (SasSwart) [reply coderd/notifications/testdata/rendered-templates/TemplateUserAccountDeleted-body.md.golden:5]: I'd be keen for a professional Wordsmith to have a look if we want.

File: coderd/users.go
Comment (SasSwart) [reply coderd/users.go:198]: > Correct me if I'm wrong, in this case, "coder" indicates the system? 

This is correct

File: coderd/notifications/testdata/rendered-templates/TemplateTemplateDeleted-body.md.golden
Comment (SasSwart) [reply coderd/notifications/testdata/rendered-templates/TemplateTemplateDeleted-body.md.golden:5]: I wondered about this, but didn't want to assume I could overwrite what was specified.
Will update.

File: coderd/notifications/testdata/rendered-templates/TemplateUserAccountDeleted-body.md.golden
Comment (dannykopping) [reply coderd/notifications/testdata/rendered-templates/TemplateUserAccountDeleted-body.md.golden:5]: Agree, we should involve @EdwardAngert.

File: coderd/users.go
Comment (dannykopping) [inline coderd/users.go:594]: As a follow-up PR, this feels like something that we should inject into the context.

**Review by dannykopping — APPROVED** [review]
LGTM modulo the existing nits.

File: coderd/notifications/testdata/rendered-templates/TemplateYourAccountSuspended-body.md.golden
Comment (SasSwart) [reply coderd/notifications/testdata/rendered-templates/TemplateYourAccountSuspended-body.md.golden]: fixed

File: coderd/notifications/notifications_test.go
Comment (SasSwart) [reply coderd/notifications/notifications_test.go]: fixed

File: coderd/database/migrations/000262_improve_notification_templates.down.sql
Comment (SasSwart) [reply coderd/database/migrations/000262_improve_notification_templates.down.sql]: fixed

File: coderd/database/migrations/000262_improve_notification_templates.down.sql
Comment (SasSwart) [reply coderd/database/migrations/000262_improve_notification_templates.down.sql]: fixed

File: coderd/database/migrations/000262_improve_notification_templates.down.sql
Comment (SasSwart) [reply coderd/database/migrations/000262_improve_notification_templates.down.sql]: fixed

File: coderd/notifications/notifications_test.go
Comment (mafredri) [inline coderd/notifications/notifications_test.go:769]: Why don't we use initiator for these fields like in other places. I feel like templates would be simplified if we didn't add the specific prefix either (created/deleted/etc.).

File: coderd/notifications/notifications_test.go
Comment (mafredri) [inline coderd/notifications/notifications_test.go:780]: Why lower case here (cmp. above)? Feels like there's a bug when looking at golden files.

File: coderd/notifications/testdata/rendered-templates/TemplateUserAccountDeleted-body.md.golden
Comment (dannykopping) [reply coderd/notifications/testdata/rendered-templates/TemplateUserAccountDeleted-body.md.golden:5]: I think we can do this in a follow-up. Let's get this merged.

**Review by mafredri — COMMENTED** [review]
Few minor things but otherwise LGTM.

**Comment by mtojek** [comment]
Re https://github.com/coder/coder/pull/14967#discussion_r1793470910:

> I wondered about this, but didn't want to assume I could overwrite what was specified.
Will update.

@SasSwart @dannykopping  Right now the result looks strange. Please take a look below at the notification I received today, it needs more love. I still don't think we need to refer to the _display name_ specifically.

Screenshot 2024-10-12 at 21 18 40

File: coderd/notifications/notifications_test.go
Comment (SasSwart) [reply coderd/notifications/notifications_test.go:780]: fixed in subsequent PR

File: coderd/notifications/notifications_test.go
Comment (SasSwart) [reply coderd/notifications/notifications_test.go:769]: fixed in subsequent PR

**Pull request merged**
